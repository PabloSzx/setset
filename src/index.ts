export * from './manager'
export * from './settings'
export * from './static'
