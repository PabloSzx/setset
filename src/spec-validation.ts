import * as Lo from 'lodash'
import { inspect } from 'util'
import {
  appendPath,
  isNamespaceSpecifier,
  isRecordSpecifier,
  renderPath,
  Specifier,
  SpecifierNamespace,
  TraversalInfo,
} from './settings'

/**
 * Validate the spec for basic invariants.
 * todo mapData mapEntryData fixup etc.
 */
export function validateSpecifier(specifier: Specifier, info: TraversalInfo) {
  if (isNamespaceSpecifier(specifier)) {
    Lo.forOwn(specifier.fields, (fieldSpecifier: Specifier, fieldName: string) => {
      validateSpecifier(fieldSpecifier, appendPath(info, fieldName))
    })
    return
  }

  if (isRecordSpecifier(specifier)) {
    validateSpecifier(specifier.entry as SpecifierNamespace, info)
    return
  }

  if (specifier.mapType !== undefined && typeof specifier.mapType !== 'function') {
    throw new Error(
      `Type mapper for setting "${renderPath(
        info
      )}" was invalid. Type mappers must be functions. Got: ${inspect(specifier.mapType)}`
    )
  }
}
