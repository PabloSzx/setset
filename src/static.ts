/**
 * This module only deals with the type level (no term level)
 * It contains the rich conditional type system powering the settings library.
 */

import { PartialDeep, Primitive } from 'type-fest'
import {
  ExcludePlainObjectOrInterface,
  ExcludeUndefined,
  Includes,
  IncludesPlainObjectOrInterface,
  IncludesRecord,
  IsSameKeys,
  KeepOptionalKeys,
  KeepRequiredKeys,
  Lookup,
  OnlyPlainObjectOrInterface,
  PlainObject,
} from './utils'

export type ExcludeShorthand<T> = OnlyPlainObjectOrInterface<T>

export type DataDefault<input> = {
  [k in keyof input]-?: IncludesPlainObjectOrInterface<input[k]> extends true
    ? DataDefault<OnlyPlainObjectOrInterface<input[k]>>
    : ExcludeUndefined<input[k]>
}

export type KeysWhereDataRequiredOrNotInData<Input, Data> = {
  [K in keyof Input]: K extends keyof Data ? (undefined extends Data[K] ? never : K) : K
}[keyof Input]

type KeysWhereDataOptionalOrNotInData<Input, Data> = {
  [K in keyof Input]: K extends keyof Data ? (undefined extends Data[K] ? K : never) : K
}[keyof Input]

type FilterInOverlappingKeys<T, U> = {
  [K in keyof T]: K extends keyof U ? K : never
}[keyof T]

type OnlyPropsInOther<T, U> = {
  [K in FilterInOverlappingKeys<T, U>]: T[K]
}

type NO_DATA_MATCH = 'NO_DATA_MATCH'

type Node<Input, Data, Key> = IncludesRecord<Lookup<Input, Key>> extends true
  ? RecordSpec<Lookup<Input, Key>, Key, Data>
  : IncludesPlainObjectOrInterface<Lookup<Input, Key>> extends true
  ? NamespaceSpec<Lookup<Input, Key>, Key, Data>
  : LeafSpec<Lookup<Input, Key>, Key, Data>

/**
 * todo
 */
// prettier-ignore
export type Spec<Input, Data> =
  (
    KeysWhereDataOptionalOrNotInData<Input, Data> extends never
      ? {}
      : { [Key in KeysWhereDataOptionalOrNotInData<Input, Data>]+?: Node<Input,Data,Key> }
  ) &
  (
    KeysWhereDataRequiredOrNotInData<Input, Data> extends never
      ? {}
      : { [Key in KeysWhereDataRequiredOrNotInData<Input, Data>]-?: Node<Input,Data,Key> }
  )

//prettier-ignore
export type NamespaceSpec<Input, Key, Data> =
  {
    fields: Spec<OnlyPlainObjectOrInterface<Input>, Lookup<Data, Key, NO_DATA_MATCH>>
  } &
  // todo jsdoc 1) when namespace has no matching data key then developer is responsible
  // todo to map the data tree over _somehow_. Impossible to know how, so return type is
  // todo any possible data. This logic is arbitrary and not guaranteed to work. You should
  // todo unit test it!
  (
    Lookup<Data, Key> extends never
      ? { mapData(input:ExcludeShorthand<Input>): PartialDeep<Data> }
      : {}
  ) &
  /**
   * If namespace is union with non-pojo type then shorthand required 
   */
  (
    ExcludeUndefined<ExcludePlainObjectOrInterface<Input>> extends never
    ? {}
    : {
        shorthand: Shorthand<ExcludeUndefined<ExcludePlainObjectOrInterface<Input>>, OnlyPlainObjectOrInterface<Input>>
      }
  ) &
  /**
   * If namespace is optional AND 1+ sub inputs are required THEN initializer required 
   *  ... if no data match THEN still required
   *  ... but if data optional THEN initial is forbidden (since we'll initialize namespace (data) to undefined)
   *  ... but if all namespace fields (input) are optional THEN initial is forbidden (b/c we can automate
   *      namespace (data) with namespace (input) field initializers)
   */
  (
    undefined extends Input
      ? {} extends KeepOptionalKeys<Input>
        ? {}
        : Includes<Lookup<Data,Key>, undefined> extends true
          ? {}
          : { initial(): KeepRequiredKeys<Exclude<Input, undefined>> }
      : {}
  )

// [1]
// If the field can be undefined it means that initial is not required.
// In most cases it probably means initial won't be supplied. However
// there are may be some odd cases where iniital is present but can
// return undefined.
// prettier-ignore
export type LeafSpec<Input, Key, Data> =
  {
    validate?: Validate<ExcludeUndefined<Input>>
    /**
     * Specify a fixup for this setting.
     *
     * A "fixup" corrects minor problems in a
     * given setting. It also provides a human readable message about what was
     * done and why.
     *
     * Return null if no fixup was needed. Return a fixup object
     * otherwise. The new value should be returned along with a list of one or
     * more messages, one for each thing that was fixed.
     */
    fixup?: Fixup<ExcludeUndefined<Input>>
  } &
  /**
   * if input is optional then initial is required
   * if input is optional and no matching data key then initial is required
   * if input is optional and matching data key optional then initial is optional
   */
  (
    undefined extends Input
      ? undefined extends Lookup<Data, Key>
        ? { initial?(): Input }
        : { initial(): ExcludeUndefined<Input> }
      : {}
  ) &
  /**
   * If there is not a direct mapping between input/data then
   * require mapping functions.
   */
  (
    // do not consider `... | undefined` b/c existence is handled specially
    NO_DATA_MATCH extends Data
      ? {} 
      :  Key extends keyof ExcludeUndefined<Data>
        ?
          ExcludeUndefined<Input> extends ExcludeUndefined<Data>[Key]
          ? {}
          : // if input key type does not match data then mapType is required
            { mapType: MapType<ExcludeUndefined<Input>, ExcludeUndefined<Data>[Key]> }
        : // if input key has no match in data then mapData is required
          { mapData: MapType<ExcludeUndefined<Input>, ExcludeUndefined<Data>> }
  )

/**
 * todo: currently assumes Record<string, object>
 */
// prettier-ignore
// todo how does no data match affect this?
export type RecordSpec<Dict, K, Data, DictEntry = Lookup<ExcludeUndefined<Dict>, string>> =
  // | { raw(input: Dict): Data[K] }
  (
    {
      /**
       * Specify the settings input for each entry in the record.
       */
      // todo how does no data match affect this?
      entry: Node<OnlyPlainObjectOrInterface<ExcludeUndefined<Dict>>, Lookup<Data,K>, string>
    }
  ) &
  // (
  //   KeysWhereDataRequired<Spec<VarEntryInput, Data[K][string]>, Data[K][string]> extends never
  //   ? {
  //       /**
  //        * Specify the record entry settings.
  //        */
  //       entryFields: Spec<VarEntryInput, Data[K][string]>
  //     }
  //   : {
  //     a: [Data, K]
  //       /**
  //        * Specify the settings that each entry will have.
  //        * 
  //        * @remarks Optional because all of the settings data is optional.
  //        */
  //       entryFields?: Spec<VarEntryInput, Data[K][string]>
  //     }
  //   ) &
    // /**
    //  * If namespace is union with non-pojo type then shorthand required
    //  */
    // (ExcludePlainObjectOrInterface<DictEntry> extends never
    //   ? {}
    //   : {
    //       // Reason for ExcludeUndefined is that a type with optional properties + index
    //       // sig will force the index sig to have undefined. This pattern seems common enough
    //       // to support here, and harmless, since, pure records shouldn't have semantic meaning of
    //       // undefined value on keys.
    //       entryShorthand(
    //         input: ExcludeUndefined<ExcludePlainObjectOrInterface<DictEntry>>
    //       ): OnlyPlainObjectOrInterface<DictEntry>
    //     }) &
    /**
     * if input is optional then initial is required
     * unless all
     */
    (
      undefined extends Dict
      ? {
          /**
           * Initialize the record with some entries. By default the record will be an empty object.
           */
          initial?(): ExcludeUndefined<Dict>
        }
      : {
          /**
           * Initialize the record with some entries. Although you require users to input a record, your initializer will still be run too, if provided.
           */
          initial?(): ExcludeUndefined<Dict>
        }
    ) &
    /**
     * if data has fields that are not present in input THEN mapData is required
     */
    (
      // todo how does no data match affect this?
      IsSameKeys<Required<Lookup<Lookup<Data,K>,string>>, ExcludeShorthand<Required<DictEntry>>> extends true
      ? {}
      : {
          // todo how does no data match affect this?
          mapEntryData(data: OnlyPropsInOther<Lookup<Lookup<Data,K>,string>, ExcludeShorthand<DictEntry>>, key: string): Lookup<Lookup<Data,K>,string>
        }
    )

/**
 * Isolated Types
 */

export type Validate<Input = Primitive> = (value: Input) => null | { messages: string[] }

export type Fixup<Input = Primitive> = (input: Input) => null | { value: Input; messages: string[] }

export type MapType<Input = Primitive, Return = Primitive> = (input: Input) => Return

export type Shorthand<Input = Primitive, Return = PlainObject> = (input: Input) => Return
