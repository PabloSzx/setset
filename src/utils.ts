import { Primitive } from 'type-fest'

/**
 * Check if curerntly in production mode defined as
 * NODE_ENV environment variable equaling "production".
 */
export function isProduction(): boolean {
  return process.env.NODE_ENV === 'production'
}

/**
 * Check if curerntly in development mode defined as
 * NODE_ENV environment variable not equaling "production".
 */
export function isDevelopment(): boolean {
  return process.env.NODE_ENV !== 'production'
}

export type Lookup<Object, Key, Fallback = never> = Key extends keyof Object ? Object[Key] : Fallback

export type IsSameKeys<A, B> = keyof A extends keyof B ? (keyof B extends keyof A ? true : false) : false

export type AnyObject = { [k: string]: any }

export declare type ExcludeUndefined<A> = A extends undefined ? never : A

export type IncludesVoid<T> = (T extends void ? true : never) extends never ? false : true

export type IncludesUndefined<T> = (T extends undefined ? true : never) extends never ? false : true

export type IsRecord<T> = string extends keyof T ? true : false

export type OnlyIndexedType<T> = string extends keyof T ? T : never

export type IncludesRecord<T> = OnlyIndexedType<Exclude<T, Primitive | void>> extends never ? false : true

export type MaybePromise<T = void> = T | Promise<T>

export type CallbackRegistrer<F> = (f: F) => void

export type SideEffector = () => MaybePromise

export type OnlyPlainObjectOrInterface<T> = T extends Function
  ? never
  : T extends RegExp
  ? never
  : T extends Date
  ? never
  : T extends any[]
  ? never
  : T extends Primitive
  ? never
  : T

export type IncludesPlainObjectOrInterface<T> = OnlyPlainObjectOrInterface<T> extends never ? false : true

export type ExcludePlainObjectOrInterface<T> = Exclude<T, OnlyPlainObjectOrInterface<T>>

/**
 * Represents a POJO. Prevents from allowing arrays and functions
 *
 * @remarks
 *
 * TypeScript interfaces will not be considered sub-types. For that, use AnyObject
 */
export type PlainObject = {
  [x: string]: Primitive | object
}

export type isPlainObject<T> = T extends PlainObject ? true : false

export type NotPlainObject = Primitive | any[] | Function

export type IncludesPlainObject<T> = Only<T, PlainObject> extends never ? false : true

export type Only<T, U> = Exclude<T, Exclude<T, U>>

export type KeepOptionalKeys<t> = {
  [k in keyof t]: undefined extends t[k] ? t[k] : never
}

export type GetRequiredKeys<T> = { [K in keyof T]: undefined extends T[K] ? never : K }[keyof T]

export type KeepRequiredKeys<t> = {
  [k in ExcludeUndefined<GetRequiredKeys<t>>]-?: undefined extends t[k] ? never : t[k]
}

export type UnknownFallback<T, U> = unknown extends T ? U : T

export type Includes<T, U> = (T extends U ? T : never) extends never ? false : true

export type ExcludePrimitive<T> = Exclude<T, Primitive>

export function assertTypeEquals<T>(expression: T): void {}

export function mergeShallow(o1: any, o2: any) {
  for (const [k, v] of Object.entries(o2)) {
    if (v !== undefined) {
      o1[k] = v
    }
  }
  return o1
}
