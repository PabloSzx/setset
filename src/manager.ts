import * as Logger from '@nexus/logger'
import * as Lo from 'lodash'
import { DataDefault, MetadataState, Spec } from '.'
import { commit, createInfo, dataFromMetadata, FixupInfo, initialize, resolve } from './settings'
import { validateSpecifier } from './spec-validation'
import { isDevelopment, PlainObject } from './utils'

const log = Logger.log.child('settings')

/**
 * todo
 */
export type Manager<Input extends PlainObject, Data extends PlainObject> = {
  reset(): Manager<Input, Data>
  change(input: Input): Manager<Input, Data>
  original(): Data
  metadata: MetadataState<Data>
  data: Data
}

export type Options = {
  /**
   * Handle fixup events.
   *
   * If your settings spec has no fixups then you can ignore this option.
   *
   * By default, fixups are logged at warning level. If you provide your own
   * function then this default behaviour will be disabled. You can retain it by
   * calling the default function passed as a second argument to your function.
   */
  onFixup?: (info: FixupInfo, originalHandler: (info: FixupInfo) => void) => void
}

/**
 *
 */
export function create<Input extends PlainObject, Data extends PlainObject = DataDefault<Input>>({
  fields,
  ...options
}: {
  fields: Spec<Input, Data>
} & Options): Manager<Input, Data> {
  log.debug('construct')
  const info = createInfo()

  if (isDevelopment()) {
    validateSpecifier({ fields }, info)
  }

  // todo we currently have to clone the given spec deeply because mapEntryData mutations the spec with shadow specifiers
  // and shodow specifiers currently break the second+ initialize run (e.g. during reset)
  // todo we didn't catch this yet with own unit test, but other nexus unit tets caught the issue

  const initial = initialize({ fields: Lo.cloneDeep(fields) }, info)
  const state = {
    data: initial.data as Data,
    original: (undefined as any) as Data, // lazy
    metadata: initial.metadata as any, // Metadata<Data>,
  }

  const api: Manager<Input, Data> = {
    data: state.data,
    metadata: state.metadata,
    change(input) {
      log.debug('change', { input })
      const newData = resolve(options, 'change', { fields }, input, state.data, state.metadata, info)
      commit({ fields: Lo.cloneDeep(fields) }, 'change', newData, state.data, state.metadata)
      return api
    },
    reset() {
      log.debug('reset')
      const initial = initialize({ fields: Lo.cloneDeep(fields) }, info)
      api.data = state.data = initial.data as any
      api.metadata = state.metadata = initial.metadata as any
      return api
    },
    original() {
      log.debug('get original')
      const original = state.original ?? dataFromMetadata(state.metadata, {})
      return original
    },
  }

  return api
}
