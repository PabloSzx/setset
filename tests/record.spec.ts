import * as S from '../src'
import { R } from './__helpers'

describe('specifiers', () => {
  it('can be omitted when input+data is optional', () => {
    const s3 = S.create<{ a?: R<{ b?: number }> }, { a?: R<{ b?: number }> }>({ fields: {} })
    s3.change({ a: { foobar: { b: 2 } } })
    expect(s3.data).toEqual({ a: { foobar: { b: 2 } } })
  })
})

it('can have their settings changed', () => {
  const s = S.create<{ a: R<{ b: number }> }>({ fields: { a: { entry: { fields: { b: {} } } } } })
  expect(s.change({ a: { foobar: { b: 2 } } }).data).toEqual({ a: { foobar: { b: 2 } } })
})

describe('initial()', () => {
  it('is accepted and is called (required record case)', () => {
    // prettier-ignore
    const s = S.create<{ a: R<{ b: number }> }>({ fields: { a: { initial: () => ({ foobar: { b: 2 } }), entry: { fields: { b: {} } } } } })
    expect(s.data).toEqual({ a: { foobar: { b: 2 } } })
    expect(s.metadata).toMatchSnapshot()
  })
  it('is accepted and is called (optional record case)', () => {
    // prettier-ignore
    const s = S.create<{ a?: R<{ b: number }> }>({ fields: { a: { initial: () => ({ foobar: {b:2}}), entry: { fields: { b: {} } } } } })
    expect(s.data).toEqual({ a: { foobar: { b: 2 } } })
    expect(s.metadata).toMatchSnapshot()
  })
  it('is omittable, defaulting to empty object (optional reocrd case)', () => {
    // todo bit odd that data could be typed as optional but it would never be...
    // prettier-ignore
    const s = S.create<{ a?: R<{ b: number }> }>({ fields: { a: { entry: { fields: { b: {} } } } } })
    expect(s.data).toEqual({ a: {} })
    expect(s.metadata).toMatchSnapshot()
  })
  describe('metadata', () => {
    it('captured immutable initial state', () => {
      const s = S.create<{ a?: R<{ b: number }> }>({ fields: { a: { entry: { fields: { b: {} } } } } })
      s.change({ a: { foobar: { b: 1 } } })
      expect(s.metadata).toMatchSnapshot()
    })
    it('captures immutable initial state even with sub-initializers', () => {
      // prettier-ignore
      const s = S.create<{ a: R<{ b?: number; c?: number }> }>({
          fields: { a: { initial: () => ({ foobar: { c: 1 } }), entry: { fields: { c: { initial: () => 100 }, b: { initial: () => 2 } } } } }
        })
      expect(s.metadata).toMatchSnapshot()
      s.change({ a: { foobar: { c: 3 } } })
      expect(s.metadata).toMatchSnapshot()
    })
  })
  describe('sub-initializers', () => {
    it('run when respective fields not given by record initializer', () => {
      const s = S.create<{ a: R<{ b?: number }> }>({
        fields: { a: { initial: () => ({ foobar: {} }), entry: { fields: { b: { initial: () => 1 } } } } },
      })
      expect(s.data).toEqual({ a: { foobar: { b: 1 } } })
    })
    it('skipped when respective fields are given by record initializer', () => {
      // prettier-ignore
      const s = S.create<{ a: R<{ b?: number }> }>({
          fields: { a: { initial: () => ({ foobar: { b: 2 } }), entry: { fields: { b: { initial: () => 1 } } } } },
        })
      expect(s.data).toEqual({ a: { foobar: { b: 2 } } })
    })
    it('data merged with data given in record initializer', () => {
      // prettier-ignore
      const s = S.create<{ a: R<{ b?: number; c: 1 }> }>({
          fields: { a: { initial: () => ({ foobar: { c: 1 } }), entry: { fields: { c: {}, b: { initial: () => 1 } } } } }
        })
      expect(s.data).toEqual({ a: { foobar: { c: 1, b: 1 } } })
    })
  })
})
describe('mapEntryType', () => {
  it.todo('required if the entry input type does not match data type')
})
describe('mapEntryData', () => {
  it('required if the entry input field name does not match any data field name; called if given', () => {
    // prettier-ignore
    const s = S.create<{ a?: R<{ a?: number }> }, { a: R<{ a: number, b: number }> }>({
        fields: { a: {  mapEntryData: (data) => ({ a: data.a, b: data.a }), entry: { fields: { a: { initial: () => 1 } } } } }
      })
    s.change({ a: { foobar: { a: 1 } } })
    expect(s.data).toEqual({ a: { foobar: { a: 1, b: 1 } } })
  })
  it('passes key name to callback as second parameter', () => {
    let keyName_
    // prettier-ignore
    const s = S.create<{ a?: R<{ a?: number }> }, { a: R<{ a: number, b: number }> }>({
        fields: { a: { mapEntryData: (data, keyName) => (keyName_ = keyName, { a: data.a, b: data.a }), entry: { fields: { a: { initial: () => 1 } } } } }
      })
    s.change({ a: { foobar: { a: 1 } } })
    expect(keyName_).toEqual('foobar')
  })
  it('runs at initialization time', () => {
    // prettier-ignore
    const s = S.create<{ a?: R<{ a?: number }> }, { a: R<{ a: number, b: number }> }>({
        fields: { a: { mapEntryData: (data) => ({ a: data.a, b: data.a }), initial: () => ({ foobar: { a: 1 } }), entry: { fields: { a: { initial: () => 1 } } } } }
      })
    expect(s.data).toEqual({ a: { foobar: { a: 1, b: 1 } } })
  })
  it('metadata considers source of contribution as being "initial" (instead of e.g. "set") and immutable as usual', () => {
    // prettier-ignore
    const s = S.create<{ a?: R<{ a?: number }> }, { a: R<{ a: number, b: number }> }>({
        fields: { a: { mapEntryData: (data) => ({ a: data.a, b: data.a }), initial: () => ({ foobar: { a: 1 } }), entry: { fields: { a: { initial: () => 1 } } } } }
      })
    const metadataAInitial = s.metadata.fields.a.initial
    expect(s.metadata).toMatchSnapshot()
    s.change({ a: { foobar: { a: 2 } } })
    expect(s.metadata.fields.a.initial).toEqual(metadataAInitial)
  })
})
describe('entryShorthand', () => {
  it.todo('can be provided to allow shorthands on entries (like namespaces)')
})
