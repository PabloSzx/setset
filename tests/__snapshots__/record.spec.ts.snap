// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial() is accepted and is called (optional record case) 1`] = `
Object {
  "fields": Object {
    "a": Object {
      "from": "initial",
      "initial": Object {
        "foobar": Object {
          "fields": Object {
            "b": Object {
              "from": "initial",
              "initial": 2,
              "type": "leaf",
              "value": 2,
            },
          },
          "type": "namespace",
        },
      },
      "type": "record",
      "value": Object {
        "foobar": Object {
          "fields": Object {
            "b": Object {
              "from": "initial",
              "initial": 2,
              "type": "leaf",
              "value": 2,
            },
          },
          "type": "namespace",
        },
      },
    },
  },
  "type": "namespace",
}
`;

exports[`initial() is accepted and is called (required record case) 1`] = `
Object {
  "fields": Object {
    "a": Object {
      "from": "initial",
      "initial": Object {
        "foobar": Object {
          "fields": Object {
            "b": Object {
              "from": "initial",
              "initial": 2,
              "type": "leaf",
              "value": 2,
            },
          },
          "type": "namespace",
        },
      },
      "type": "record",
      "value": Object {
        "foobar": Object {
          "fields": Object {
            "b": Object {
              "from": "initial",
              "initial": 2,
              "type": "leaf",
              "value": 2,
            },
          },
          "type": "namespace",
        },
      },
    },
  },
  "type": "namespace",
}
`;

exports[`initial() metadata captures immutable initial state even with sub-initializers 1`] = `
Object {
  "fields": Object {
    "a": Object {
      "from": "initial",
      "initial": Object {
        "foobar": Object {
          "fields": Object {
            "b": Object {
              "from": "initial",
              "initial": 2,
              "type": "leaf",
              "value": 2,
            },
            "c": Object {
              "from": "initial",
              "initial": 1,
              "type": "leaf",
              "value": 1,
            },
          },
          "type": "namespace",
        },
      },
      "type": "record",
      "value": Object {
        "foobar": Object {
          "fields": Object {
            "b": Object {
              "from": "initial",
              "initial": 2,
              "type": "leaf",
              "value": 2,
            },
            "c": Object {
              "from": "initial",
              "initial": 1,
              "type": "leaf",
              "value": 1,
            },
          },
          "type": "namespace",
        },
      },
    },
  },
  "type": "namespace",
}
`;

exports[`initial() metadata captures immutable initial state even with sub-initializers 2`] = `
Object {
  "fields": Object {
    "a": Object {
      "from": "initial",
      "initial": Object {
        "foobar": Object {
          "fields": Object {
            "b": Object {
              "from": "initial",
              "initial": 2,
              "type": "leaf",
              "value": 2,
            },
            "c": Object {
              "from": "initial",
              "initial": 1,
              "type": "leaf",
              "value": 1,
            },
          },
          "type": "namespace",
        },
      },
      "type": "record",
      "value": Object {
        "foobar": Object {
          "fields": Object {
            "b": Object {
              "from": "initial",
              "initial": 2,
              "type": "leaf",
              "value": 2,
            },
            "c": Object {
              "from": "set",
              "initial": 1,
              "type": "leaf",
              "value": 3,
            },
          },
          "type": "namespace",
        },
      },
    },
  },
  "type": "namespace",
}
`;

exports[`mapEntryData metadata considers source of contribution as being "initial" (instead of e.g. "set") and immutable as usual 1`] = `
Object {
  "fields": Object {
    "a": Object {
      "from": "initial",
      "initial": Object {
        "foobar": Object {
          "fields": Object {
            "a": Object {
              "from": "initial",
              "initial": 1,
              "type": "leaf",
              "value": 1,
            },
            "b": Object {
              "from": "initial",
              "initial": 1,
              "isShadow": true,
              "type": "leaf",
              "value": 1,
            },
          },
          "type": "namespace",
        },
      },
      "type": "record",
      "value": Object {
        "foobar": Object {
          "fields": Object {
            "a": Object {
              "from": "initial",
              "initial": 1,
              "type": "leaf",
              "value": 1,
            },
            "b": Object {
              "from": "initial",
              "initial": 1,
              "isShadow": true,
              "type": "leaf",
              "value": 1,
            },
          },
          "type": "namespace",
        },
      },
    },
  },
  "type": "namespace",
}
`;
